File: backend/src/fmp-api.ts
// Financial Modeling Prep API 클라이언트
import axios from 'axios'

const FMP_API_KEY = 'Nz122fIiH3KWDx8UVBdQFL8a5NU9lRhc'
const FMP_BASE_URL = 'https://financialmodelingprep.com/api/v3'

interface CompanyProfile {
  symbol: string
  companyName: string
  price: number
  exchange: string
  exchangeShortName: string
  industry: string
  sector: string
  isActivelyTrading: boolean
}

interface StockQuote {
  symbol: string
  price: number
  changesPercentage: number
  change: number
  dayLow: number
  dayHigh: number
  volume: number
  marketCap: number
  exchange: string
}

interface HistoricalPrice {
  date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

interface IntradayPrice {
  date: string
  minute: string
  label: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export class FMPApi {
  /**
   * 회사명으로 심볼 검색 (FMP Name Search API) - 정확도 개선
   */
  async searchByName(query: string): Promise<any[]> {
    try {
      // 1단계: Name Search API (회사명 정확 검색)
      const nameUrl = `${FMP_BASE_URL}/search-name?query=${encodeURIComponent(query)}&apikey=${FMP_API_KEY}`
      const nameResponse = await axios.get(nameUrl, { timeout: 10000 })
      
      console.log(`🔍 FMP 회사명 검색: "${query}"`)
      console.log(`   Name Search 결과: ${nameResponse.data.length}개`)
      
      // 나스닥만 필터링 및 정확도 점수 계산
      const nasdaqResults = nameResponse.data
        .filter((item: any) => {
          const exchange = item.exchangeShortName || item.stockExchange
          return exchange === 'NASDAQ' || exchange === 'NMS' || exchange === 'NGM' || exchange === 'NCM'
        })
        .map((item: any) => {
          // 회사명 매칭 점수 계산 (정확도)
          const companyName = (item.name || item.companyName || '').toLowerCase()
          const queryLower = query.toLowerCase()
          
          let score = 0
          
          // 완전 일치
          if (companyName === queryLower) {
            score = 100
          }
          // 시작 일치
          else if (companyName.startsWith(queryLower)) {
            score = 90
          }
          // 포함 (단어 경계)
          else if (new RegExp(`\\b${queryLower}\\b`).test(companyName)) {
            score = 80
          }
          // 포함 (일반)
          else if (companyName.includes(queryLower)) {
            score = 70
          }
          // 부분 일치 (각 단어)
          else {
            const queryWords = queryLower.split(/\s+/)
            const matchedWords = queryWords.filter(word => companyName.includes(word))
            score = (matchedWords.length / queryWords.length) * 60
          }
          
          return { ...item, matchScore: score }
        })
        .filter(item => item.matchScore >= 60) // 60점 이상만
        .sort((a, b) => b.matchScore - a.matchScore) // 점수 높은 순
      
      if (nasdaqResults.length > 0) {
        console.log(`   ✓ 나스닥 종목 발견:`)
        nasdaqResults.slice(0, 3).forEach(item => {
          console.log(`      ${item.symbol} - ${item.name || item.companyName} (점수: ${item.matchScore})`)
        })
      } else {
        console.log(`   ✗ 나스닥 종목 없음`)
      }
      
      return nasdaqResults
    } catch (error: any) {
      console.error('FMP name search error:', error.message)
      return []
    }
  }

  /**
   * 티커 심볼 검색 (Symbol Search API)
   */
  async searchSymbol(query: string): Promise<any[]> {
    try {
      const url = `${FMP_BASE_URL}/search?query=${encodeURIComponent(query)}&apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 10000 })
      
      // 나스닥만 필터링
      const nasdaqResults = response.data.filter((item: any) => {
        const exchange = item.exchangeShortName || item.stockExchange
        return exchange === 'NASDAQ' || exchange === 'NMS' || exchange === 'NGM' || exchange === 'NCM'
      })
      
      return nasdaqResults
    } catch (error: any) {
      console.error('FMP symbol search error:', error.message)
      return []
    }
  }

  /**
   * 티커 검증 (정확한 티커인지 확인)
   */
  async validateSymbol(symbol: string): Promise<{ valid: boolean; profile: any | null }> {
    try {
      // 1. Profile API로 티커 존재 여부 확인
      const profile = await this.getCompanyProfile(symbol)
      
      if (!profile) {
        console.log(`   ✗ "${symbol}" 는 유효하지 않거나 나스닥이 아님`)
        return { valid: false, profile: null }
      }
      
      // 2. 활성 거래 중인지 확인
      if (!profile.isActivelyTrading) {
        console.log(`   ✗ "${symbol}" 는 거래 중단됨`)
        return { valid: false, profile: null }
      }
      
      console.log(`   ✓ "${symbol}" 검증 성공: ${profile.companyName}`)
      return { valid: true, profile }
    } catch (error: any) {
      console.error(`티커 검증 실패 ${symbol}:`, error.message)
      return { valid: false, profile: null }
    }
  }

  /**
   * 티커의 회사 프로필 조회
   */
  async getCompanyProfile(symbol: string): Promise<CompanyProfile | null> {
    try {
      const url = `${FMP_BASE_URL}/profile/${symbol}?apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 10000 })
      
      if (response.data && response.data.length > 0) {
        const profile = response.data[0]
        
        // 나스닥 거래소인지 확인
        const isNasdaq = profile.exchangeShortName === 'NASDAQ' || 
                        profile.exchangeShortName === 'NMS'
        
        if (!isNasdaq) {
          return null
        }
        
        return {
          symbol: profile.symbol,
          companyName: profile.companyName,
          price: profile.price,
          exchange: profile.exchange,
          exchangeShortName: profile.exchangeShortName,
          industry: profile.industry,
          sector: profile.sector,
          isActivelyTrading: profile.isActivelyTrading
        }
      }
      
      return null
    } catch (error) {
      console.error(`FMP profile error for ${symbol}:`, error)
      return null
    }
  }

  /**
   * 실시간 주가 조회 (단일)
   */
  async getQuote(symbol: string): Promise<StockQuote | null> {
    try {
      const url = `${FMP_BASE_URL}/quote/${symbol}?apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 10000 })
      
      if (response.data && response.data.length > 0) {
        const quote = response.data[0]
        return {
          symbol: quote.symbol,
          price: quote.price,
          changesPercentage: quote.changesPercentage,
          change: quote.change,
          dayLow: quote.dayLow,
          dayHigh: quote.dayHigh,
          volume: quote.volume,
          marketCap: quote.marketCap,
          exchange: quote.exchange
        }
      }
      
      return null
    } catch (error) {
      console.error(`FMP quote error for ${symbol}:`, error)
      return null
    }
  }

  /**
   * 실시간 주가 조회 (다중) - 여러 종목 한번에
   */
  async getQuotes(symbols: string[]): Promise<StockQuote[]> {
    try {
      if (symbols.length === 0) return []

      // FMP는 쉼표로 구분된 심볼 지원
      const symbolsParam = symbols.join(',')
      const url = `${FMP_BASE_URL}/quote/${symbolsParam}?apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 15000 })
      
      if (response.data && Array.isArray(response.data)) {
        return response.data.map((quote: any) => ({
          symbol: quote.symbol,
          price: quote.price,
          changesPercentage: quote.changesPercentage,
          change: quote.change,
          dayLow: quote.dayLow,
          dayHigh: quote.dayHigh,
          volume: quote.volume,
          marketCap: quote.marketCap,
          exchange: quote.exchange
        }))
      }
      
      return []
    } catch (error) {
      console.error(`FMP quotes error:`, error)
      return []
    }
  }

  /**
   * 티커가 나스닥에 상장되어 있는지 확인
   */
  async isNasdaqListed(symbol: string): Promise<boolean> {
    const profile = await this.getCompanyProfile(symbol)
    return profile !== null && profile.isActivelyTrading
  }

  /**
   * 뉴스 텍스트에서 회사명으로 티커 찾기
   */
  async findTickerFromText(text: string): Promise<string | null> {
    // 텍스트에서 가능한 회사명 추출
    const possibleCompanies = this.extractCompanyNames(text)
    
    for (const companyName of possibleCompanies) {
      const results = await this.searchSymbol(companyName)
      
      if (results.length > 0) {
        // 가장 관련성 높은 결과 (첫 번째) 반환
        const topResult = results[0]
        console.log(`✓ Found ticker: ${topResult.symbol} for "${companyName}"`)
        return topResult.symbol
      }
    }
    
    return null
  }

  /**
   * 텍스트에서 회사명 추출 (간단한 휴리스틱)
   */
  private extractCompanyNames(text: string): string[] {
    const names: string[] = []
    
    // 대문자로 시작하는 단어들 (2-4 단어)
    const regex = /\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+){1,3})\b/g
    let match
    
    while ((match = regex.exec(text)) !== null) {
      const name = match[1]
      // 일반적인 단어 제외
      if (!['The', 'This', 'That', 'These', 'Those', 'October', 'November'].includes(name)) {
        names.push(name)
      }
    }
    
    return names.slice(0, 5) // 최대 5개만
  }

  /**
   * 일별 히스토리컬 데이터 조회 (최대 1년)
   */
  async getHistoricalPrices(ticker: string, days: number = 30): Promise<HistoricalPrice[]> {
    try {
      // 조정되지 않은 실제 가격 사용 (스플릿 반영 안 함)
      const url = `${FMP_BASE_URL}/historical-price-full/${ticker}?apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 10000 })
      
      if (response.data?.historical) {
        const data = response.data.historical.slice(0, days)
        console.log(`📊 FMP Historical API 응답 (${ticker}):`, data.length, '개')
        if (data.length > 0) {
          console.log('   최신 데이터:', data[0])
        }
        return data
      }
      return []
    } catch (error) {
      console.error(`히스토리컬 데이터 조회 실패 ${ticker}:`, error)
      return []
    }
  }

  /**
   * 실시간 인트라데이 데이터 조회 (FMP Stable API - 유료 플랜)
   */
  async getIntradayPrices(ticker: string, interval: string = '5min'): Promise<IntradayPrice[]> {
    try {
      // interval: '1min', '3min' (X), '5min', '15min', '30min', '1hour', '4hour'
      // FMP는 3분봉을 지원하지 않으므로 15분봉으로 대체
      const apiInterval = interval === '3min' ? '15min' : interval
      
      // FMP Stable API 엔드포인트 (유료 플랜 - 실시간 데이터)
      const url = `https://financialmodelingprep.com/stable/historical-chart/${apiInterval}?symbol=${ticker}&apikey=${FMP_API_KEY}`
      const response = await axios.get(url, { timeout: 15000 })
      
      if (Array.isArray(response.data)) {
        console.log(`📊 FMP 실시간 ${apiInterval} 데이터 (${ticker}):`, response.data.length, '개')
        if (response.data.length > 0) {
          console.log(`   최신: ${response.data[0].date} - $${response.data[0].close}`)
        }
        
        // 데이터 제한 (성능 고려)
        // 1분봉: 최대 5일치 (1950개)
        // 5분봉: 최대 30일치 (2340개)
        // 15분봉: 최대 90일치 (3600개)
        const limit = interval === '1min' ? 1950
                    : interval === '5min' ? 2340
                    : interval === '15min' ? 3600
                    : interval === '30min' ? 4800
                    : interval === '1hour' ? 7200
                    : 2000
        
        return response.data.slice(0, limit)
      }
      return []
    } catch (error: any) {
      // 에러 발생 시 구버전 API로 폴백
      console.log(`⚠️ Stable API 실패, 구버전 API로 폴백: ${ticker}`)
      try {
        const fallbackUrl = `${FMP_BASE_URL}/historical-chart/${interval}/${ticker}?apikey=${FMP_API_KEY}`
        const fallbackResponse = await axios.get(fallbackUrl, { timeout: 10000 })
        if (Array.isArray(fallbackResponse.data)) {
          return fallbackResponse.data.slice(0, 2000)
        }
      } catch (fallbackError) {
        console.error(`인트라데이 데이터 조회 완전 실패 ${ticker}:`, fallbackError)
      }
      return []
    }
  }

  /**
   * 여러 티커 검증 (배치)
   */
  async validateTickers(symbols: string[]): Promise<Record<string, boolean>> {
    const results: Record<string, boolean> = {}
    
    // 병렬로 검증 (최대 5개씩)
    const chunks = []
    for (let i = 0; i < symbols.length; i += 5) {
      chunks.push(symbols.slice(i, i + 5))
    }
    
    for (const chunk of chunks) {
      const promises = chunk.map(async (symbol) => {
        const isValid = await this.isNasdaqListed(symbol)
        results[symbol] = isValid
      })
      
      await Promise.all(promises)
    }
    
    return results
  }

  /**
   * 현재가 조회 (KIS API 대체/보조용)
   */
  async getCurrentPrice(symbol: string): Promise<number | null> {
    const quote = await this.getQuote(symbol)
    return quote ? quote.price : null
  }
}

File: scanner/utils/fmp_api.py
# utils/fmp_api.py
"""
FMP (Financial Modeling Prep) API 호출 유틸리티
최신 /stable/ 엔드포인트 사용
"""
import os
import requests
from dotenv import load_dotenv
from typing import Dict, List, Any

# .env 로드
load_dotenv()

FMP_KEY = os.getenv("FMP_API_KEY")
BASE = "https://financialmodelingprep.com/stable"

def _get(path: str, params: Dict[str, Any] | None = None) -> Any:
    """
    FMP API GET 요청 공통 함수
    """
    if FMP_KEY is None:
        raise RuntimeError("FMP_API_KEY가 .env 파일에 설정되지 않았습니다")
    
    params = params or {}
    
    # 파라미터를 URL 쿼리 스트링으로 변환
    query_parts = []
    for k, v in params.items():
        query_parts.append(f"{k}={v}")
    query_parts.append(f"apikey={FMP_KEY}")
    query_string = "&".join(query_parts)
    
    url = f"{BASE}{path}?{query_string}"
    
    try:
        r = requests.get(url, timeout=15)
        # 에러 상태일 때만 로그 출력
        if r.status_code != 200:
            print(f"[API ERROR] {path} - Status: {r.status_code}")
        r.raise_for_status()
        return r.json()
    except requests.exceptions.RequestException as e:
        print(f"[ERROR] FMP API 호출 실패: {url}")
        print(f"[ERROR] {e}")
        return None

def get_profile(symbol: str) -> List[Dict] | None:
    """
    종목 프로필 조회 (시가총액, 가격 등)
    새 API: /stable/quote?symbol=AAPL (profile API 대체)
    
    응답 형식:
    [
        {
            "symbol": "AAPL",
            "name": "Apple Inc.",
            "price": 232.8,
            "marketCap": 3500823120000,  // <- marketCap (mktCap 아님!)
            "exchange": "NASDAQ",
            ...
        }
    ]
    """
    return _get("/quote", {"symbol": symbol})

def get_quote(symbol: str) -> List[Dict] | None:
    """
    종목 현재가 조회
    새 API: /stable/quote?symbol=AAPL
    """
    return _get("/quote", {"symbol": symbol})

def get_batch_quotes(symbols: List[str]) -> List[Dict] | None:
    """
    여러 종목 현재가 일괄 조회 (정규장용 배치 API)
    새 API: /stable/batch-quote?symbols=AAPL,MSFT,GOOGL
    
    symbols: 종목 리스트 (최대 50개 권장)
    
    반환 형식:
    [
        {
            "symbol": "AAPL",
            "name": "Apple Inc.",
            "price": 270.14,
            "changePercentage": 0.037,
            "change": 0.1,
            "volume": 40361476,
            "marketCap": 3991696696000,
            ...
        },
        ...
    ]
    """
    if not symbols:
        return []
    
    # 심볼을 쉼표로 연결
    symbols_str = ",".join(symbols[:50])  # 최대 50개씩
    return _get("/batch-quote", {"symbols": symbols_str})

def get_batch_aftermarket_quotes(symbols: List[str]) -> List[Dict] | None:
    """
    여러 종목 애프터마켓/프리마켓 현재가 일괄 조회
    새 API: /stable/batch-aftermarket-trade?symbols=AAPL,MSFT,GOOGL
    
    symbols: 종목 리스트 (최대 50개 권장)
    
    반환 형식:
    [
        {
            "symbol": "AAPL",
            "price": 269.83,
            "tradeSize": null,
            "timestamp": 1762390796000
        },
        ...
    ]
    """
    if not symbols:
        return []
    
    # 심볼을 쉼표로 연결
    symbols_str = ",".join(symbols[:50])  # 최대 50개씩
    return _get("/batch-aftermarket-trade", {"symbols": symbols_str})

def get_hist_daily(symbol: str, days: int = 400) -> Dict | None:
    """
    일봉 히스토리 조회
    새 API: /stable/historical-price-eod/full?symbol=AAPL
    
    반환 형식:
    {
        "symbol": "AAPL",
        "historical": [
            {
                "date": "2025-02-04",
                "open": 227.2,
                "high": 233.13,
                "low": 226.65,
                "close": 232.8,
                "volume": 44489128,
                "change": 5.6,
                "changePercent": 2.46479,
                "vwap": 230.86
            }
        ]
    }
    """
    result = _get("/historical-price-eod/full", {"symbol": symbol})
    
    # 결과가 리스트 형식이면 {"symbol": symbol, "historical": result} 형태로 변환
    if result and isinstance(result, list):
        return {
            "symbol": symbol,
            "historical": result[:days] if days else result
        }
    return result

def get_hist_1min(symbol: str, minutes: int = 390*5) -> List[Dict] | None:
    """
    1분봉 히스토리 조회
    새 API: /stable/historical-chart/1min?symbol=AAPL
    
    정규장(RTH) 기준 최근 minutes 분 데이터
    
    반환 형식:
    [
        {
            "date": "2025-02-04 15:59:00",
            "open": 233.01,
            "low": 232.72,
            "high": 233.13,
            "close": 232.79,
            "volume": 720121
        }
    ]
    """
    result = _get("/historical-chart/1min", {"symbol": symbol})
    
    # minutes 제한 적용
    if result and isinstance(result, list) and minutes:
        return result[:minutes]
    return result

def get_hist_5min(symbol: str, bars: int = 2000) -> List[Dict] | None:
    """
    5분봉 히스토리 조회
    새 API: /stable/historical-chart/5min?symbol=AAPL
    
    bars: 반환할 최대 봉 수
    """
    result = _get("/historical-chart/5min", {"symbol": symbol})
    
    if result and isinstance(result, list) and bars:
        return result[:bars]
    return result

def get_hist_15min(symbol: str, bars: int = 2000) -> List[Dict] | None:
    """
    15분봉 히스토리 조회
    새 API: /stable/historical-chart/15min?symbol=AAPL
    
    bars: 반환할 최대 봉 수
    """
    result = _get("/historical-chart/15min", {"symbol": symbol})
    
    if result and isinstance(result, list) and bars:
        return result[:bars]
    return result

if __name__ == "__main__":
    # 테스트
    print("=== FMP API 테스트 ===")
    
    # 프로필 조회 테스트
    profile = get_profile("AAPL")
    if profile:
        print(f"\nAAPL 프로필: {profile[0].get('name', 'N/A')}")
        print(f"시가총액: ${profile[0].get('marketCap', 0):,}")
    
    # 현재가 조회 테스트
    quote = get_quote("AAPL")
    if quote:
        print(f"\nAAPL 현재가: ${quote[0].get('price', 0)}")
    
    print("\n테스트 완료")

